packages/clerk-js/src/ui/Components.tsx

```diff
@@ -149,7 +160,21 @@ const Components = (props: ComponentsProps) => {
   const { signInModal, signUpModal, userProfileModal, organizationProfileModal, createOrganizationModal, nodes } =
     state;
 
+  // TODO: rename these
+  const urlClerkState = getClerkQueryParam('__clerk_state') ?? '';
+  const decodedRedirectParams = urlClerkState ? JSON.parse(atob(urlClerkState)) : null;
+
   useSafeLayoutEffect(() => {
+    if (urlClerkState && decodedRedirectParams && decodedRedirectParams?.modal) {
+      setState(s => ({
+        ...s,
+        [componentNodes[snakeToCamel(decodedRedirectParams.basePath)]]: {},
```

Comment:
❓ Component paths are dynamic. For example, the <UserProfile/> is located at `/user` by default but users can change it to anything. So the mapping between paths and components should be dynamic. 

Can we replace that basePath resolution with a property that says which component needs to be loaded on redirect? 

@nikosdouvlis WDTY?

Reply:
A couple of things here.

1. As said before, reading + decoded + cleaning the state param should done with helper functions

2. We are only using this logic if the component was opened as a modal. So instead of always appending the state and checking the `modal` key, let's only append the state when it's needed in the first place. So if we have serialized state, we always perform the action dictated by the state (in this case, we open a modal)

3. We shouldn't have to do `options.startPath` . `options` is the public API of ClerkProvider and type changes to this are user-facing changes that we don't want 

4. I'm not sure what `componentNodes` represents or what `        [componentNodes[snakeToCamel(decodedRedirectParams.basePath)]]: {},` actually does... Can you please elaborate?

Reply:
Generally, it might be easier to append the `componentName` and the current path when the flow starts and just append it to the startPath, eg: 
```
              startPath={'/organization' + serializedState.path}
```
(this is just an example as closing the modal afterward should also clear the path so next time it opens, it opens as it normally would)